cli/lib/webpack/service.js

```diff
@@ -0,0 +1,226 @@
+
+	ensureDaemonIsRunning() {
+		return new Promise((resolve, reject) => {
+			this.logger.debug('Connecting to appcd ...');
+			this.client.connect({ startDaemon: true })
+				.once('connected', () => {
+					this.logger.info(`🖥  Web UI ready on http://localhost:1732${this.basePath}/web`);
+					resolve();
+				})
+				.once('error', reject);
+		});
+	}
+
+	ensureWebpackIsRunning() {
+		const options = {
+			identifier: this.jobIdentifier,
+			watch: this.shouldWatch,
+			project: this.projectSettings,
+			build: this.buildSettings
+		};
+
+		return new Promise((resolve, reject) => {
+			if (this.forceRebuild) {
+				this.startWebpack(options).then(resolve).catch(reject);
+				return;
+			}
+
+			this.client
+				.request({
+					path: `${this.basePath}/status/${options.identifier}`,
+				})
+				.once('response', response => {
+					this.webpackStatus = response;
+					if (this.webpackStatus.state === 'stopped' || this.webpackStatus.state === 'error') {
+						this.logger.debug('Webpack not running, starting new Webpack build.');
+						this.startWebpack(options).then(resolve).catch(reject);
+					} else {
+						if (!this.validateOptions(response.options, options)) {
+							this.logger.debug('Build options changed, restarting Webpack build.');
+							this.startWebpack(options).then(resolve).catch(reject);
+							return;
+						}
+						if (this.webpackStatus.state === 'ready') {
+							this.builder.tiSymbols = response.tiSymbols;
+						}
+						return resolve();
+					}
+				})
+				.once('error', error => {
+					if (error.status === 404) {
+						this.startWebpack(options).then(resolve).catch(reject);
+						return;
+					}
+
+					reject(error);
+				});
+		});
+	}
+
+	async waitForCompilationComplete() {
+		if (this.webpackStatus.state === 'ready') {
+			this.logger.info('Build is up-to-date and ready!');
+			if (!this.shouldWatch) {
+				this.client.disconnect();
+			}
+			return Promise.resolve();
+		}
+
+		this.logger.info('Waiting for compilation to complete ...');
+		await this.subcribeToWebpackStatusChanges();
+
+		return new Promise((resolve, reject) => {
+			const handler = e => {
+				if (e.state === 'ready' || e.state === 'error') {
+					this.webpackStatusEmitter.off('status', handler);
+
+					if (e.state === 'ready') {
+						clearTimeout(buildTimeout);
+						if (!this.shouldWatch) {
+							this.client.disconnect();
+						}
+						return resolve();
+					} else {
+						return reject(new Error('Webpack compilation failed.'));
+					}
+				}
+			};
+			this.webpackStatusEmitter.on('status', handler);
+
+			// @todo what to do in this case? Give useful tips!
+			const buildTimeout = setTimeout(() => reject(new Error('Did not receive success message from Webpack appcd plugin within 120s, aborting...')), 120000);
+		});
+	}
+
+	startWebpack(options) {
+		this.logger.debug(`Starting Webpack with options: ${JSON.stringify(options)}`);
+		this.webpackStatus = { state: 'building' };
+		return new Promise((resolve, reject) => {
```

Comment:
really not a fan of all this hand-rolling of Promises, but I'm not sure there's a nice way to move to `async`/`await` with events like this.

Reply:
Yeah, me neither, but i don't think there is a nice way around this since `appcd-client` is event based.

cc @cb1kenobi What do you think about having an alternative promise based API for simple requests like these? Always having to write request like the following can get quite repetitive.

```js
client
  .request({
    path: '/appcd/status'
  })
  .once('response', onResponse)
  .once('error', onError);
```

I could imagine a promise based API with `get(path): Promise` and `post(path, data): Promise`, for example.